fix(batcher): altda parallel submitted blobs respect strict holocene order

samlaf opened the pull request:
See [upstreamed PR](https://github.com/ethereum-optimism/optimism/pull/13169) for more details. OP doesn't have time to review and upstream our PRs atm so we have to maintain this fork instead.

The recent holocene strict ordering rules broke the altda concurrent blob submission flow because concurrent requests to the da-server could be received out of order, and hence would be dropped by the op-node's derivation pipeline with strict ordering rules.

Solution TLDR: cache the altda commitments inside their respective channel, and add logic to pull them back in order.

This PR looks big but most of it is tests. The main Solution logic is quite small, and contained in this one commit: [3cbe31f](https://github.com/ethereum-optimism/optimism/pull/13169/commits/3cbe31fed3423229f5dbe17e29f1948de290488c)

It also contains a fix for the recently introduced sync_actions data race ([0a8bbe4](https://github.com/ethereum-optimism/optimism/pull/13169/commits/0a8bbe4f8b57cbbdf79f2db3ea1503dddf3ca60a)), but I think this will be subsumed by https://github.com/ethereum-optimism/optimism/pull/13330

Otherwise the added tests are:
1. e2e test that has the fakeDAServer return blobs out of order: [bebf7bb](https://github.com/ethereum-optimism/optimism/pull/13169/commits/bebf7bbe063d089a55f3cc28c139bff7e571941b)
2. unit tests for unhappy path failure cases: [a6c2830](https://github.com/ethereum-optimism/optimism/pull/13169/commits/a6c28304f3883d43e760640436588099790a95b3)
3. unit test that triggers channelTimeouts using the [failpoint](https://github.com/pingcap/failpoint) library: [a43ed4d](https://github.com/ethereum-optimism/optimism/pull/13169/commits/a43ed4df97d33a666a763da2095542556d668aef)
this one is more contentious as it requires some source code manipulation (b/c golang doesnt have macros) codegen. Updated the tooling (mise, just dependencies) to make it reproducible, and also using a separate GOCACHE for it, so should be very self-contained. But still left it as a separate commit so that we can easily remove it if you guys don't want it (even though I think you should, and should probably use it to test channelTimeouts for non-altda codepaths!)

samlaf committed:
test(e2e): new altda e2e test for concurrent blob submissions to maintain new holocene strict ordering rules

samlaf committed:
test(batcher): add altda unit tests for unhappy failure cases (except channel timeout)

test(batcher): fix flaky driver tests + speed them up

test(batcher): robustify batcher driver altda tests

samlaf committed:
fix(batcher): altda concurrent blob responses are reordered to respect holocene strict ordering rules

litt3 commented on op-alt-da/damock.go:
```diff
@@ -16,22 +18,53 @@ import (
 }
 
+// NewCountingGenericCommitmentMockDAClient creates a MockDAClient that uses counting commitments.
+// It's commitments are big-endian encoded uint16s of 0, 1, 2, etc. instead of actual hash or altda-layer related commitments.
```
nit: "It's" -> "Its"

samlaf replied:
[ac57dc0](https://github.com/Layr-Labs/optimism/pull/21/commits/ac57dc0fb4a2c93eabd78575ed2bb743c72333ee)

litt3 commented on op-alt-da/damock.go:
```diff
@@ -40,12 +73,42 @@ func (c *MockDAClient) GetInput(ctx context.Context, key CommitmentData) ([]byte
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.log.Debug("Deleting data", "key", key)
+	err := c.store.Delete(key)
```
`Delete` doesn't specify in its contract that an error will be returned if the key isn't present. In fact, the `memorydb` impl used in this mock is a noop if key isn't present

Presumably, the `if` below which decrements `StoreCount` needs to be conditional on the key existing in the store prior to delete

samlaf replied:
[6211a9e](https://github.com/Layr-Labs/optimism/pull/21/commits/6211a9eeadfc9e2513f59f1d9862d0f531d9b3d8)

litt3 commented on op-batcher/batcher/channel.go:
```diff
@@ -38,21 +49,60 @@ func newChannel(log log.Logger, metr metrics.Metricer, cfg ChannelConfig, rollup
 	}
 }
 
+func (s *channel) CacheAltDACommitment(txData txData, commitment altda.CommitmentData) {
```
nit: missing doc for exported method

samlaf replied:
[ac57dc0](https://github.com/Layr-Labs/optimism/pull/21/commits/ac57dc0fb4a2c93eabd78575ed2bb743c72333ee)

litt3 commented on op-batcher/batcher/channel.go:
```diff
@@ -124,6 +189,28 @@ func (c *channel) ID() derive.ChannelID {
 	return c.channelBuilder.ID()
 }
 
+// NextAltDACommitment checks if it has already receives the altDA commitment
```
"if it has already receives" -> "if it has already *received*"

samlaf replied:
[ac57dc0](https://github.com/Layr-Labs/optimism/pull/21/commits/ac57dc0fb4a2c93eabd78575ed2bb743c72333ee)

litt3 commented on op-batcher/batcher/channel.go:
```diff
@@ -124,6 +189,28 @@ func (c *channel) ID() derive.ChannelID {
+		if len(txData.frames) == 0 {
+			panic("expected txData to have frames")
+		}
+		// update altDANextFrame to the first frame of the next txData
```
`altDANextFrame` an old name? Perhaps this comment should talk about `altDAFrameCursor` instead

samlaf replied:
[ac57dc0](https://github.com/Layr-Labs/optimism/pull/21/commits/ac57dc0fb4a2c93eabd78575ed2bb743c72333ee)

litt3 commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -90,6 +91,34 @@ func (s *channelManager) pendingBlocks() int {
 	return s.blocks.Len() - s.blockCursor
 }
 
+func (s *channelManager) CacheAltDACommitment(txData txData, commitment altda.CommitmentData) {
```
nit: missing docs for `CacheAltDACommitment` and `AltDASubmissionFailed`

samlaf replied:
[ac57dc0](https://github.com/Layr-Labs/optimism/pull/21/commits/ac57dc0fb4a2c93eabd78575ed2bb743c72333ee)

samlaf committed:
docs: fix typos and add documentation comments for some batcher public methods

samlaf committed:
test(op-alt-da): fix MockDAClient.DeleteData decrement semantic

samlaf committed:
chore(batcher): move channel failover behavior from TxFailed to AltDASubmissionFailed

The failover logic in the failover feature commits was aded on TxFailed, as the separate function AltDASubmissionFailed didn't exist yet. This change makes it much cleaner as a tx having failed cannot lead to a failover given... since that would come from an ethereum issue.

samlaf committed:
fix(batcher): bug in sendTransaction

samlaf committed:
chore(test-logger): fix test logger.Crit which wasn't getting flushed

samlaf committed:
test(batcher): fix altDASetup w new channel config

DaType was added, so needed to change config to use DaTypeAltDA

samlaf commented on op-batcher/justfile:
```diff
@@ -17,7 +17,13 @@ clean:
+test: test-failpoint
+    just go_test "./..."
+
+# Failpoint tests rely on https://github.com/pingcap/failpoint, which are identified by a _FailpointTest suffix to the test name.
+# We run them using a separate GOCACHE because they inject code into the compiled binaries, which could interfere with other tests or normal builds.
+test-failpoint:
+    GOCACHE=/tmp/failpoint-cache just go_test "./batcher/..." "-toolexec" "$(mise which failpoint-toolexec)" "-run" ".*_FailpointTest$"
```
@litt3 would like to chat tomorrow about this specific test framework needed for a single test. Considering just removing that test, would like to discuss.

samlaf replied:
Conclusion: https://github.com/Layr-Labs/optimism/pull/21#issuecomment-2721688570

samlaf commented:
After syncing with @litt3 , decided to keep the failpoint test for batcher-timeouts out of this PR. It brought in too much complexity and external dependency for such a small benefit. Would also make merging upstream harder in the future given that op has explicitly told me they would not merge a PR with the failpoint dependency.
Moved that commit to its own archive branch in case we need to refer to it in the future: https://github.com/Layr-Labs/optimism/commit/4b79c981a13bf096ae2984634d976956fbbfddff

samlaf commented on op-batcher/batcher/driver_test.go:
```diff
@@ -114,5 +123,218 @@ func TestBatchSubmitter_SafeL1Origin_FailsToResolveRollupClient(t *testing.T) {
+}
+
+func TestBatchSubmitter_AltDA_FailureCase3_ChannelTimeout(t *testing.T) {
+	// TODO: implement this test
```
TODO: mention why we can't implement this (dont want to use failpoint dependency)

samlaf replied:
[387c2e5](https://github.com/Layr-Labs/optimism/pull/21/commits/387c2e5e87170ff44d5e7b9dd2f28baa0397b6a1)

samlaf commented on op-batcher/readme.md:
```diff
@@ -38,23 +38,32 @@ In the happy path, the batcher periodically:
 
-### Tx Failed
+> TODO: is this true?? `waitNodeSync()` seems to only wait for the node to sync to the L1 tip of the batcher's L1 node, or an older block where the last batcher tx was included, but not wait for inflight txs to be ingested.
+
+#### Tx Failed
 When a Tx fails, an asynchronous receipts handler is triggered. The channel from whence the Tx's frames came has its `frameCursor` rewound, so that all the frames can be resubmitted in order.
 
-### Channel Times Out
+> TODO: there might be an issue with this simple logic. See https://github.com/ethereum-optimism/optimism/issues/13283
+
+#### Channel Times Out
 When a Tx is confirmed, an asynchronous receipts handler is triggered. We only update the batcher's state if the channel timed out on chain. In that case, the `blockCursor` is rewound to the first block added to that channel, and the channel queue is cleared out. This allows the batcher to start fresh building a new channel starting from the same block -- it does not need to refetch blocks from t ...
 
+#### AltDA Submission Fails
+
+> TODO: describe how the batcher handles this.
```
bunch of TODOs here

samlaf replied:
[197b065](https://github.com/Layr-Labs/optimism/pull/21/commits/197b06504796607f54a270bbcd7a8513fe41cad3)

litt3 commented on op-alt-da/damgr.go:
```diff
@@ -78,8 +78,12 @@ type DA struct {
+func NewAltDA(log log.Logger, cli CLIConfig, cfg Config, metrics Metricer) (*DA, error) {
+	daClient, err := cli.NewDAClient()
+	if err != nil {
+		return nil, err
```
nit: wrap error

samlaf replied:
[e47b398](https://github.com/Layr-Labs/optimism/pull/21/commits/e47b3984a5ce3abae5075d78d936b58d86081c58)

samlaf committed:
style: wrap errors

samlaf committed:
style(damock): dont explicitly initialize to 0

We make use default values instead

samlaf commented on op-alt-da/damock.go:
```diff
@@ -16,22 +18,53 @@ import (
+	StoreCount             int
+	log                    log.Logger
+	dropEveryNthPut        uint // 0 means nothing gets dropped, 1 means every put errors, etc.
+	setInputRequestCount   uint // number of put requests received, irrespective of whether they were successful
```
The DAClient has 2 methods GetInput and SetInput, so `setInputRequestCount` is the number of times `SetInput` was called. Set is the name op uses for "write" here, which is unfortunate but don't want to change it.

samlaf committed:
docs(batcher-test): document why channel timeout test is left unimplemented

samlaf committed:
docs(batcher): fix todos in batcher readme

samlaf commented on op-alt-da/cli.go:
```diff
@@ -102,8 +102,12 @@ func (c CLIConfig) Check() error {
+func (c CLIConfig) NewDAClient() (*DAClient, error) {
+	err := c.Check()
+	if err != nil {
+		return nil, err
```
[e47b398](https://github.com/Layr-Labs/optimism/pull/21/commits/e47b3984a5ce3abae5075d78d936b58d86081c58)

samlaf commented on op-alt-da/damock.go:
```diff
@@ -16,22 +18,53 @@ import (
+		CommitmentType:         GenericCommitmentType,
+		GenericCommitmentCount: 0,
+		store:                  memorydb.New(),
+		StoreCount:             0,
```
[0b4a90e](https://github.com/Layr-Labs/optimism/pull/21/commits/0b4a90e3b64221d7ba0f3a91bf88a04fa702a2bf)

samlaf commented on op-alt-da/damock.go:
```diff
@@ -16,22 +18,53 @@ import (
 	return &MockDAClient{
 		CommitmentType: Keccak256CommitmentType,
 		store:          memorydb.New(),
+		StoreCount:     0,
```
[0b4a90e](https://github.com/Layr-Labs/optimism/pull/21/commits/0b4a90e3b64221d7ba0f3a91bf88a04fa702a2bf)

samlaf committed:
chore: make lint-go-fix

litt3 commented on op-batcher/readme.md:
```diff
@@ -50,19 +50,17 @@ In each channel, the `frameCursor` tracks the next unsent frame.
 #### AltDA Submission Fails
 
-> TODO: describe how the batcher handles this.
+When an AltDA submission fails, the frames that pushed back into their respective channel, and will be retried in the next tick. If the da-server returns a 503 HTTP error, then failover to ethDA-calldata is triggered for that specific channel. Each channel will independently always first try to submit to EigenDA.
```
Did you mean to write:

"the frames that pushed back" -> "the frames get pushed back" ?

samlaf replied:
thanks for catching all of these! [89ac378](https://github.com/Layr-Labs/optimism/pull/21/commits/89ac378da3ef86cd2c8ae4656de7a9a1a2ed06f7)

samlaf committed:
docs(batcher): fix readme typo

samlaf committed:
Merge branch 'eigenda-develop' into fix--batcher-altda-concurrent-blobs-respect-strict-holocene-order-rebase
